### PR TITLE
Adding repeat instance to the model for repeat

### LINKED ISSUE #182: Don't always include jr:template for repeats
Currently `jr:template` is always added to repeats but it is not generally what users want. Typically users don't provide defaults for questions in the repeat.

I think the ideal would be to only include `jr:template` if one or more of the questions in the repeat has a default. Another option would be to add something like `template=yes|no` to the `parameters` column.

### PR DESCRIPTION
**Pull request by nribeka**
Fix #182.

This will add instance for repeat section.

### CODE REVIEW

**Comment by nribeka** [comment]
@lognaturel 2 failed tests. The tests is counting number of elements in the model which obviously going to fail because we're adding new elements to the model. I'm just wondering if this is what we're looking for.

File: pyxform/section.py [hunk 1]
```diff
@@ -40,12 +42,30 @@ def xml_instance(self, **kwargs):
...
             elif isinstance(child, ExternalInstance):
                 continue
+            else:
+                if isinstance(child, RepeatingSection) and not append_template:
+                    append_template = not append_template
+                    repeating_template = child.generate_repeating_template()
+                result.appendChild(child.xml_instance(append_template=append_template))
+            if append_template and repeating_template:
+                append_template = not append_template
+                result.insertBefore(repeating_template, result._get_lastChild())
```
Comment (nribeka) [inline pyxform/section.py]: ```result._get_lastChild()``` vs ```result.childNodes[-1]```?

File: pyxform/section.py [hunk 2]
```diff
@@ -104,11 +120,11 @@ def xml_control(self):
 
     # I'm anal about matching function signatures when overriding a function,
```
Comment (lognaturel) [inline pyxform/section.py]: This comment is no longer relevant, right?

Comment (nribeka) [reply pyxform/section.py]: Remove it?

File: pyxform/tests_v1/test_repeat_template.py [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+        repeat_b_template = '<repeat_b jr:template="">'
+        self.assertEqual(1, survey_xml.count(repeat_b_template))
+
+        section_instance = "<section>"
+        self.assertEqual(1, survey_xml.count(section_instance))
+        repeat_a_instance = "<repeat_a>"
+        self.assertEqual(1, survey_xml.count(repeat_a_instance))
+        repeat_b_instance = "<repeat_b>"
+        self.assertEqual(1, survey_xml.count(repeat_b_instance))
+
```
[the remark this answers is not in the document]
Comment (nribeka) [reply pyxform/tests_v1/test_repeat_template.py]: Ah I didn't know about that. I just assume to use the equality assertion 👯‍♂

**Comment by lognaturel** [comment]
I talked to @seadowg about this who pointed out that it's quite likely form designers are currently using repeats in Collect to represent 0 or more of a thing (rather than 1 or more). For example, if the desire is to collect information about children or siblings, it's quite likely that the flow of getting prompted initially about whether or not to add a repeat and potentially saying no is actually the right flow.

I don't know exactly what to do about this now. Marking as `needs discussion`.

**Comment by nribeka** [comment]
This would also need input from @MartijnR.

**Comment by MartijnR** [comment]
Enketo creates the first repeat automatically (and has been since 2012). In the past users have expressed a desire to have 0, but I believe this has been addressed sufficiently by introducing support for repeat-count and allowing the first repeat to be deleted manually (which initially wasn't possible). Since then, here have been no issues reported. If it is important to (be able to) start with 0, I agree we need to address that though.

**Comment by craigappl** [comment]
@ukanga , note that we need to review the comments from today's session. [Link](https://github.com/XLSForm/pyxform/issues/182#issuecomment-541444640)

**Comment by nribeka** [comment]
@lognaturel or @ukanga do you guys aware of way to do warning assertions in the v1? There's an old test that assert the warning, but it's the old unit test.

File: pyxform/section.py (hunk 2 above)
Comment (lognaturel) [reply pyxform/section.py]: This was about overriding `xml_instance` and the name has now changed, right? So yes, the comment should be removed. It's also looking to me like neither this method nor `generate_repeating_template` need to have a `**kwargs` parameter, is that right?

File: pyxform/xls2json.py [hunk 4]
```diff
@@ -859,6 +860,20 @@ def workbook_to_json(
...
+                if (
+                    control_type == constants.REPEAT
+                    and not repeat_behavior_warning_added
+                ):
+                    warnings.append(
+                        "Repeat behavior has changed. Previously, some clients like "
+                        "ODK Collect prompted users to add the first repeat. Now, "
+                        "the user will only be prompted to add repeats after the first "
+                        "one. Representing 0 repetitions will require changing the form "
+                        "design. Read more at http://xlsform.org/en/#repeats."
```
Comment (lognaturel) [inline pyxform/xls2json.py]: You can now change this to http://xlsform.org#representing-zero-repeats

**Comment by lognaturel** [comment]
I think you can see an example of verifying warnings with v1 tests in the `test_language_warnings` test.

**Comment by nribeka** [comment]
@lognaturel I just checked that unit tests and the warning testing will not work on this use case because the warning is generated when we parse the xls to json, but the v1 tests doesn't use the same function. It's using custom function to convert the md to json.

File: pyxform/xls2json.py (hunk 4 above)
Comment (nribeka) [reply pyxform/xls2json.py]: Updated the link to the above.

**Pull request merged**